### ISSUE TITLE
Command options to turn off optimizations

### ISSUE THREAD

**Issue by tomgoddard** [tlitfin/lmi4boltz#6]
I added some flags to the LMI4Boltz code to control which optimizations are enabled --use_cpu_memory and --inplace_operations.  The --use_cpu_memory option enables copying tensors from CUDA GPU memory to CPU memory and the --inplace_operations allows modifying tensors in place.  The default is false (don't optimize) for both.  I also added an --aggressive_chunking option that sets the other chunking options as described in the LMI4Boltz preprint (chunk_size_transition_z = 32, chunk_size_tri_attn = 64, triangle_mult_gate_nchunks = 4).  Also I set environment variable PYTORCH_CUDA_ALLOC_CONF = "expandable_segments:True" in the boltz predict code if it is not already set.  Also I updated LMI4Boltz to use the current jwohlwend main branch code (most recent commit Sept 8, 2025) instead of the June 8, 2025 code that the LMI4Boltz code was using.  These changes are in my RBVI (Resource for Biocomputing, Visualization and Informatics) Boltz fork, branch name low_mem:

https://github.com/RBVI/boltz/tree/low_mem#

If you would like to update the LMI4Boltz repository I could submit a pull request, or you could clone my repository and make the update from that.

I had a few reasons for making these changes.  I wanted to try LMI4Boltz on Mac and the existing code uses the tensor.cuda() method for moving tensors between CPU and CUDA memory which raises an error on systems without CUDA.  So on Mac I run without using the --use_cpu_memory flag.  I considered making the CPU/GPU memory copying work with Mac MPS (metal performance shaders), but since Mac uses unified memory for GPU and CPU I don't think this would reduce memory use.   In tests on Mac with just the --inplace_operations the memory reduction appeared to be only about 10% (using PDB 8sa0 sequence and ligands).  The expandable segments Torch setting has no effect on Mac since it is CUDA specific.  I also wanted to use LMI4Boltz on Windows and Linux in CPU-only mode but I have not yet tested that.  I would expect only modest memory reduction from --inplace_operations similar to Mac GPU.

I added --inplace_operations to control the use of inplace tensor modifications because those cause incorrect gradient back-propagation in training.  In order to allow the low memory optimizations to be merged into the main Boltz source (jwohlwend) it would of course be necessary to preserve correct training behavior.

I added the --aggressive_chunking option just for convenience.  It only sets the 3 other chunking options to the values described in the LMI4Boltz preprint.

The above changes modified about 1000 lines of code, quite messy.  I wonder if this is a sane approach to memory use optimization.  It seems like the in-place operations are a general kind of optimization that Torch should be capable of without modifying code, basically a compiler optimization.  But I have not investigated if Torch can do that.  It could be difficult because the optimizer would have to determine that no other code is using the original tensor values.  Also the transfer of seldom used tensors from GPU to CPU memory seems like a general Torch optimization that could be done without modifying Boltz code.  Maybe the hugging face Accelerator module or DeepSpeed can do this kind of thing.  Of course, human judgement on which tensors to move may be better than a generic solution.  At any rate, I think it makes sense to investigate these generic Torch memory optimization approaches given how much clutter the LMI4Boltz optimizations add to the Boltz code.
